refactor: idcc using metadata file provided by API

XavierJp opened the pull request:
references #303
references https://github.com/etalab/annuaire-entreprises-search-infra/issues/218

johangirod commented on pages/divers/[slug].tsx:
```diff
@@ -48,6 +49,8 @@ export const getServerSideProps: GetServerSideProps = postServerSideProps(
 
     const uniteLegale = await getUniteLegaleFromSlug(slug, { isBot });
 
+    uniteLegale.conventionsCollectives = await getIdccWithMetadata(uniteLegale);
```
Je ne suis pas sûr que ce soit une bonne chose de muter le modèle pour ajouter les metadata. Je préfèrerais créer une ressource séparée (`conventionCollective`), et la passer explicitement aux composants qui en ont besoin. C'est une plus grosse réécriture, mais cela a de multiples avantages : 

- On peut faire de la gestion d'erreur « classique » dessus (`APINotResponding`, `DataSection`, ...)
- Le modèle est généré de la même manière partout, pas de données différentes en fonction des routes appelées. Meilleure maintenabilité, prédictibilité et performances.

XavierJp replied:
J'aime beaucoup ta remarque.

johangirod reviewed (commented):
Je ne sais pas à quel point ça vaut le coup d'implémenter les changements proposés dans cette review, ou juste carrément de passer à l'implémentation qui fetch les métadonnées au build. 

Cela répondrait également aux problèmes soulevés et l'implémentation serait plus simple.

XavierJp commented on models/conventions-collectives-list.ts:
```diff
@@ -0,0 +1,64 @@
+      });
+    }
+  };
+}
```
Avec la variable private et le singleton, si on réussi une fois, la liste est mise en cache dans `_idccMetadata`
Avec les erreurs Sentry on va pouvoir surveiller le comportement

XavierJp commented:
@johangirod merci pour tes retours, j'ai utilisé une `<DataSection />` pour gérer les cas de `APINotRespondingError`et les cas de `NotFound`.
Comme suggéré, j'ai découpé en une ressource séparée qui porte la notion de API Not Responding. Bien plus propre, merci !

johangirod commented on models/conventions-collectives-list.ts:
```diff
@@ -0,0 +1,64 @@
+
+export const getCCMetadata = async (cc: IConventionsCollectives) => {
+  if (Object.keys(cc).length === 0) {
+    return APINotRespondingFactory(EAdministration.MTPEI, 404);
```
Ça m'a fait bizarre de voir cette erreur ici. En fait, on utilise une 404 pour encoder une exception métier, qui est « cette unité légale n'a pas de de convention collective ». Ce n'est pas que le MTPEI a répondu une 404...

Il y a des cas où c'est OK de ne pas avoir de convention collective, si la structure n'a pas de salariés par exemple.

Après, je comprends le besoin initial qui était d'afficher le `notFound` du `DataSection` dans ce cas. Mais je pose alors la question : est-ce vraiment une 404 ? Je pense qu'il faudrait ne pas le traiter comme une exception, mais comme un cas métier standard, et avoir la condition sur l'existence d'idccs pour l'unité légale dans le corps de la vue.

Sauf si quelque chose m'échappe :p 

(vu que c'est très métier, je switche en français)

XavierJp replied:
Alors, je dirais que c'est bien une 404 car il existe un vrai code pour les conventions collective non renseignées : 
`"9999":"___Sans convention collective___"` tandis que ici on a bien aucune CC trouvée dans le fichier

johangirod reviewed (approved):
C'est beaucoup mieux avec une ressource séparée, le code est plus simple à comprendre :)

Petite interrogation (non résolue) sur l'endroit où mettre la logique de cache. Est-ce le model (comme actuellement) ou le client ?

XavierJp commented:
Assez d'accord, il faut déporter la mise en cache dans le client

XavierJp commented:
@johangirod 
- I moved cache logic to the client
- I removed the 404 as it feels -indeed- clumsy and it is unnecessary as tabs is hidden when no CC is provided
